File: src/DataScience/MATLAB/example1.m
function example1
% Abstract Factory represented as a struct of related constructors.
% Select the family once, then create all related products from it.

darkFactory = createFactory(@darkButton, @darkCheckbox);
lightFactory = createFactory(@lightButton, @lightCheckbox);

createUIComponents(darkFactory);
createUIComponents(lightFactory);
end

function factory = createFactory(buttonConstructor, checkboxConstructor)
factory = struct( ...
    'createButton', buttonConstructor, ...
    'createCheckbox', checkboxConstructor);
end

function createUIComponents(factory)
buttonConstructor = factory.createButton;
checkboxConstructor = factory.createCheckbox;
buttonConstructor();
checkboxConstructor();
end

function darkButton
fprintf('Dark Button\n');
end

function lightButton
fprintf('Light Button\n');
end

function darkCheckbox
fprintf('Dark Checkbox\n');
end

function lightCheckbox
fprintf('Light Checkbox\n');
end

File: src/DataScience/Julia/example1.jl
abstract type Button end
abstract type Checkbox end

struct DarkButton <: Button end
struct LightButton <: Button end

struct DarkCheckbox <: Checkbox end
struct LightCheckbox <: Checkbox end

render(b::DarkButton) = println("Dark Button")
render(b::LightButton) = println("Light Button")
render(c::DarkCheckbox) = println("Dark Checkbox")
render(c::LightCheckbox) = println("Light Checkbox")

abstract type UIFactory end

struct DarkFactory <: UIFactory end
struct LightFactory <: UIFactory end

create_button(::DarkFactory) = DarkButton()
create_button(::LightFactory) = LightButton()

create_checkbox(::DarkFactory) = DarkCheckbox()
create_checkbox(::LightFactory) = LightCheckbox()

function create_ui_components(factory::UIFactory)
    button = create_button(factory)
    checkbox = create_checkbox(factory)
    render(button)
    render(checkbox)
end

# Usage
create_ui_components(DarkFactory())
create_ui_components(LightFactory())
